hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java

```diff
@@ -681,35 +685,36 @@ public AbfsRestOperation append(final String path, final byte[] buffer,
         abfsUriQueryBuilder, cachedSasToken);
 
     final URL url = createRequestUrl(path, abfsUriQueryBuilder.toString());
-    final AbfsRestOperation op = new AbfsRestOperation(
-        AbfsRestOperationType.Append,
-        this,
-        HTTP_METHOD_PUT,
-        url,
-        requestHeaders,
-        buffer,
-        reqParams.getoffset(),
-        reqParams.getLength(),
-        sasTokenForReuse);
+    final AbfsRestOperation op = getAbfsRestOperationForAppend(AbfsRestOperationType.Append,
+        HTTP_METHOD_PUT, url, requestHeaders, buffer, reqParams.getoffset(),
+        reqParams.getLength(), sasTokenForReuse);
     try {
       op.execute(tracingContext);
     } catch (AzureBlobFileSystemException e) {
+      // If the http response code indicates a user error we retry
+      // the same append request with expect header disabled.
+      // When "100-continue" header is enabled but a non Http 100 response comes,
+      // JDK fails to provide all response headers.
```

Comment:
There is no rfc that specifies this behaviour but in case of expect failing with any non http 100 response, the response message is set to null. That's why we mentioned that it might fail to provide all response headers.

Reply:
Today in some places, devs have taken dependency on the response message being sent by the server for the further processing in code. In case if server doesn't send a response of 100-continue, client has not sent the payload and hence the response message may be null in some cases. So to make sure that backward compatibility is not hindered, we are retrying the request failed with user errors without expect header being enabled.